### PR TITLE
Skip redundant LoggerContext reset

### PR DESCRIPTION
**Pull request by mches**
Skips a redundant call to `LoggerContext#reset()`.

* Prior to Logback 1.5.7, calling `#stop()` on _any_ `LoggerContext` implicitly called `#reset()`. On the other hand, since Logback 1.5.7, only calling `#stop()` on a _started_ `LoggerContext` implicitly calls `#reset()`. Going forward, given the next version of Spring Boot will not be binary or source compatible with Logback 1.5.6 and earlier, it's only necessary to call `#reset()` on a _stopped_ `LoggerContext`.

### CODE REVIEW

File: spring-boot-project/spring-boot/src/test/java/org/springframework/boot/logging/logback/LogbackLoggingSystemTests.java [hunk 1]
```diff
@@ -93,8 +94,9 @@
  * @author Scott Frederick
  * @author Jonatan Ivanov
  * @author Moritz Halbritter
+ * @author Mark Chesney
  */
-@ExtendWith(OutputCaptureExtension.class)
+@ExtendWith({ MockitoExtension.class, OutputCaptureExtension.class })
```
Comment (philwebb) [inline spring-boot-project/spring-boot/src/test/java/org/springframework/boot/logging/logback/LogbackLoggingSystemTests.java:99]: Is this change accidental? It doesn't seem related to the other changes?

File: spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackConfigurator.java [hunk 2]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2012-2023 the original author or authors.
+ * Copyright 2012-2024 the original author or authors.
```
Comment (philwebb) [inline spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackConfigurator.java:2]: There doesn't seem to be any changes in this file so the copyright header should not be updated.

**Review by philwebb — COMMENTED** [review]
Thanks for the PR. I've added a couple of issues to review.

File: spring-boot-project/spring-boot/src/test/java/org/springframework/boot/logging/logback/LogbackLoggingSystemTests.java (hunk 1 above)
Comment (mches) [reply spring-boot-project/spring-boot/src/test/java/org/springframework/boot/logging/logback/LogbackLoggingSystemTests.java:99]: This appeared in #41885, which was reverted.

Without this, I ran into [this issue](https://github.com/mockito/mockito/issues/1379#issuecomment-1967392526) in my environment, which failed the tests that use `Mockito.mock` within this class for being unable to delete `mockitoboot*.jar` from the JUnit temp directory. 

With the change, the test is less finicky on Windows.

That said, I can take it out if required.

**Comment by philwebb** [comment]
> Going forward, given the next version of Spring Boot will not be binary or source compatible with Logback 1.5.6 and earlier

Is this the case? Is Logback 1.5.7 not binary compatible with Logback 1.5.6?

File: spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackConfigurator.java (hunk 2 above)
Comment (mches) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackConfigurator.java:2]: This file was updated in #41885, which was reverted, but updated again without updating the years. So I feel it's accurate, but I'm happy to take it out if required.

**Comment by mches** [comment]
Thanks for the review 🙇‍♂️

**Comment by mches** [comment]
> > Going forward, given the next version of Spring Boot will not be binary or source compatible with Logback 1.5.6 and earlier
> 
> 
> 
> Is this the case? Is Logback 1.5.7 not binary compatible with Logback 1.5.6?

Yes, because the return type of `LoggerContext#getConfigurationLock()` changed from `Object` to `ReentrantLock`. It's source compatible, but not binary compatible.

**Comment by snicoll** [comment]
So this is in addition of the [actual upgrade](https://github.com/spring-projects/spring-boot/issues/41887) to Logback then?

File: spring-boot-project/spring-boot/src/test/java/org/springframework/boot/logging/logback/LogbackLoggingSystemTests.java (hunk 1 above)
Comment (wilkinsona) [reply spring-boot-project/spring-boot/src/test/java/org/springframework/boot/logging/logback/LogbackLoggingSystemTests.java:99]: I don't think it's needed. Regardless, it isn't related to this change so I think it should be taken out. https://github.com/spring-projects/spring-boot/issues/41905 has hopefully helped here.

File: spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackConfigurator.java (hunk 2 above)
Comment (wilkinsona) [reply spring-boot-project/spring-boot/src/main/java/org/springframework/boot/logging/logback/LogbackConfigurator.java:2]: I should have bumped the year as part of https://github.com/spring-projects/spring-boot/commit/dd2ed5f12d23b15ff0055e366ab6da8e95218596. That should now be done separately as it isn't related to this change so please take it out of this PR.

**Review by wilkinsona — CHANGES_REQUESTED** [review]

**Comment by mches** [comment]
> So this is in addition of the [actual upgrade](https://github.com/spring-projects/spring-boot/issues/41887) to Logback then?

Yes, I thought this made sense to do after reviewing dd2ed5f.

**Comment by mches** [comment]
Thank you for the review. All feedback has been implemented.

**Comment by wilkinsona** [comment]
I think we should leave this as-is. Arguably, calling both `stop()` and then `reset()` has been unnecessary for some time as in 1.5.6 and earlier versions a call to `stop()` will also call `reset()` but it hasn't caused any problems.

In 1.5.7, `stop()` will only call `reset()` if the context is started. A context is started in its constructor so, if it isn't started, it must have already been stopped and, therefore, must have already been reset. I think that means that we could, perhaps, do nothing at all if the context isn't started. It's hard to justify making a change as there doesn't appear to be any benefit yet there is some risk in making the change. I'm going to close this one, for now at least. We can reconsider in the future and perhaps do nothing at all if the context isn't started. Thanks anyway, @mches.
